src/lib.rs

```diff
@@ -306,8 +306,11 @@ impl WalkDir {
     /// }
     /// WalkDir::new("foo").sort_by(Some(string_compare));
     /// ```
-    pub fn sort_by(mut self, sorter: Option<fn(&OsString,&OsString) -> std::cmp::Ordering>) -> Self {
-        self.opts.sorter = sorter;
+    pub fn sort_by<F: FnMut(&OsString,&OsString) -> std::cmp::Ordering + 'static>(mut self, cmp: Option<F>) -> Self {
```

Comment:
Could we get rid of `Option` here? It can default to `None` internally, but it seems fine to force the creation of a new `WalkDir` if the caller wants to add and then drop sorting.